src/rust/engine/graph/src/entry.rs

```diff
@@ -0,0 +1,542 @@
+/// have) changed. When a dependent consumes a dependency at a particular generation, that
+/// generation is recorded on the consuming edge, and can later used to determine whether the
+/// inputs to a node have changed.
+///
+/// Unlike the RunToken (which is incremented whenever a node re-runs), the Generation is only
+/// incremented when the output of a node has changed.
+///
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub(crate) struct Generation(u32);
+
+impl Generation {
+  fn initial() -> Generation {
+    Generation(0)
+  }
+
+  fn next(&self) -> Generation {
+    Generation(self.0 + 1)
+  }
+}
+
+pub(crate) enum EntryState<N: Node> {
+  // A node that has either been explicitly cleared, or has not yet started Running. In this state
+  // there is no need for a dirty bit because the RunToken is either in its initial state, or has
+  // been explicitly incremented when the node was cleared.
+  //
+  // The previous_result value is _not_ a valid value for this Entry: rather, it is preserved in
+  // order to compute the generation value for this Node by comparing it to the new result the next
+  // time the Node runs.
+  NotStarted {
+    run_token: RunToken,
+    generation: Generation,
+    previous_result: Option<Result<N::Item, N::Error>>,
+  },
+  // A node that is running. A running node that has been marked dirty re-runs rather than
+  // completing.
+  //
+  // The `previous_result` value for a Running node is not a valid value. See NotStarted.
+  Running {
+    run_token: RunToken,
+    generation: Generation,
+    start_time: Instant,
+    waiters: Vec<oneshot::Sender<Result<(N::Item, Generation), N::Error>>>,
+    previous_result: Option<Result<N::Item, N::Error>>,
+    dirty: bool,
+  },
+  // A node that has completed, and then possibly been marked dirty. Because marking a node
+  // dirty does not eagerly re-execute any logic, it will stay this way until a caller moves it
+  // back to Running.
+  Completed {
+    run_token: RunToken,
+    generation: Generation,
+    result: Result<N::Item, N::Error>,
+    dep_generations: Vec<Generation>,
+    dirty: bool,
+  },
+}
+
+impl<N: Node> EntryState<N> {
+  fn initial() -> EntryState<N> {
+    EntryState::NotStarted {
+      run_token: RunToken::initial(),
+      generation: Generation::initial(),
+      previous_result: None,
+    }
+  }
+}
+
+///
+/// Because there are guaranteed to be more edges than nodes in Graphs, we mark cyclic
+/// dependencies via a wrapper around the Node (rather than adding a byte to every
+/// valid edge).
+///
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub(crate) enum EntryKey<N: Node> {
+  Valid(N),
+  Cyclic(N),
+}
+
+impl<N: Node> EntryKey<N> {
+  pub(crate) fn content(&self) -> &N {
+    match self {
+      &EntryKey::Valid(ref v) => v,
+      &EntryKey::Cyclic(ref v) => v,
+    }
+  }
+}
+
+///
+/// An Entry and its adjacencies.
+///
+#[derive(Clone)]
```

Comment:
Er, actually... you're now cloning the `Node` for every `Graph::get`. That's maybe not ideal? Unfortunately I have never found a smoking gun pointing to cloning `Node`s being expensive.